tools/config/__init__.py

```diff
@@ -610,24 +610,75 @@ def _get_mem_specs(self, memories, cmsis_part, exception_text):
+        rom_memories = ['IROM1', 'PROGRAM_FLASH', 'IROM2']
         cmsis_part = self._get_cmsis_part()
-        rom_start, rom_size = self._get_mem_specs(
-            ["IROM1", "PROGRAM_FLASH"],
-            cmsis_part,
-            "Not enough information in CMSIS packs to build a bootloader "
-            "project"
-        )
-        rom_start = int(getattr(self.target, "mbed_rom_start", False) or rom_start, 0)
-        rom_size = int(getattr(self.target, "mbed_rom_size", False) or rom_size, 0)
-        return (rom_start, rom_size)
+        for memory in rom_memories:
+            try:
+                rom_start, rom_size = self._get_mem_specs(
+                    memory.split(),
+                    cmsis_part,
+                    "Not enough information in CMSIS packs to build a bootloader "
+                    "project"
+                )
+                rom_start = int(rom_start, 0)
+                rom_size = int(rom_size, 0)
+                if not all_regions:
+                    return rom_start, rom_size
+                if memory=='IROM1' or memory=='PROGRAM_FLASH':
+                    memory = 'ROM'
+                else:
+                    rom_active_memories += 1
+                    memory = 'ROM' + str(rom_active_memories)
+                rom_regions[memory] = [rom_start, rom_size]
+            except ConfigException:
+                continue         
+        return rom_regions
+
+    def get_all_ram_regions(self):
+        """Get information of all available ram regions in the form of dictionary 
+        {Region: [start_addr, size]}"""
+        # Override rom_start/rom_size
+        #
+        # This is usually done for a target which:
+        # 1. Doesn't support CMSIS pack, or
+        # 2. Supports TrustZone and user needs to change its flash partition
+        ram_regions = {}
+        # Counter to keep track of RAM memories supported by target
+        ram_active_memories = 0
+        ram_memories = ['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
+                        'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER']
+        cmsis_part = self._get_cmsis_part()
+        for memory in ram_memories:
+            try:
+                ram_start, ram_size = self._get_mem_specs(
+                    memory.split(),
+                    cmsis_part,
+                    "Not enough information in CMSIS packs to build a bootloader "
+                    "project"
+                )
+                ram_start = int(ram_start, 0)
+                ram_size = int(ram_size, 0)
+                if (memory == 'IRAM1' or memory == 'SRAM_OC' or \
+                    memory == 'SRAM_UPPER') and (not self.has_ram_regions):
+                    memory = 'RAM'
+                else:
+                    ram_active_memories += 1
+                    memory = 'RAM' + str(ram_active_memories)
+                ram_regions[memory] = [ram_start, ram_size]
+            except ConfigException:
+                continue         
```

Comment:
(Sorry if I'm missing background info on this stuff, just coming up to speed on this stuff).

Are most of the regions specified in `ram_memories` expected to be present? If not, it seems a bit funny to loop through all the possible memory regions and just catching the exception. Looking at this code here: 

https://github.com/ARMmbed/mbed-os/blob/abdfcc0579f432b3c8450463b15cc7c5b4d4e7db/tools/config/__init__.py#L606-L607

Looks like it's just checking for keys in the dictionary `cmsis_part`. You could cut down a bit on the looping by doing a bit of set logic on the keys:

```python
ram_memories = set(['IRAM1', 'IRAM2', 'IRAM3', 'IRAM4', 'SRAM_OC', \
                             'SRAM_ITC', 'SRAM_DTC', 'SRAM_UPPER', 'SRAM_LOWER'])
present_memories = set(cmsis_part['memory'].keys())
valid_memories = ram_memories.intersect(present_memories)
for memory in valid_memories:
    # resume existing code but without the need for the try/except
```

Reply:
Great suggestion! I should have paid more attention to this. Thanks :)